texar/data/data/multi_aligned_data.py

```diff
@@ -0,0 +1,701 @@
+                                     "be shared only when vocab is shared.")
+                emb = embs[emb_shr]
+            else:
+                emb = None
+                emb_file = hparams_i["embedding_init"]["file"]
+                if emb_file and emb_file != "":
+                    emb = Embedding(vocabs[i].token_to_id_map_py,
+                                    hparams_i["embedding_init"])
+            embs.append(emb)
+
+        return embs
+
+    def process(self, raw_example: Tuple[Union[str, Dict[str, Any]], ...]) \
+            -> Tuple[Union[List[str], int, float, Dict[str, Any]], ...]:
+        dataset_hparams = self.hparams.datasets
+        processed_examples = []
+        for i, raw_example_i in enumerate(raw_example):
+            if _is_text_data(dataset_hparams[i]["data_type"]):
+                delimiter = dataset_hparams[i]["delimiter"]
+                raw_tokens = raw_example_i.split(delimiter)  # type: ignore
+                max_seq_length = dataset_hparams[i]["max_seq_length"]
+                if (max_seq_length is not None and
+                        len(raw_tokens) > max_seq_length):
+                    if dataset_hparams[i]["length_filter_mode"] is \
+                            _LengthFilterMode.TRUNC.value:
+                        raw_tokens = raw_tokens[:max_seq_length]
+
+                if (dataset_hparams[i].bos_token is not None and
+                        dataset_hparams[i].bos_token != ''):
+                    raw_tokens.insert(0, dataset_hparams[i].bos_token)
+                if (dataset_hparams[i].eos_token is not None and
+                        dataset_hparams[i].eos_token != ''):
+                    raw_tokens.append(dataset_hparams[i].eos_token)
+                transforms = self._other_transforms[i]
+                if len(transforms) > 0:
+                    for transform in transforms:
+                        raw_tokens = transform(raw_tokens)
+                processed_examples.append(raw_tokens)
+            elif _is_scalar_data(dataset_hparams[i]["data_type"]):
+                scalar_ex: Union[str, int, float] = str(raw_example_i)
+                data_type = dataset_hparams[i]["data_type"]
+                if data_type == "int":
+                    scalar_ex = int(scalar_ex)
+                elif data_type == "float":
+                    scalar_ex = float(scalar_ex)
+                transforms = self._other_transforms[i]
+                if len(transforms) > 0:
+                    for transform in transforms:
+                        scalar_ex = transform(scalar_ex)
+                processed_examples.append(scalar_ex)
+            elif _is_record_data(dataset_hparams[i]["data_type"]):
+                example: Dict[str, Any] = raw_example_i  # type: ignore
+                # for some reasons even after the not None check here, mypy
+                # thinks self._dataset_features[i] may be None, hence adding
+                # ignore to the lines below
```

Comment:
I think this is because `self._dataset_features[i]` calls `__getitem__` of `_dataset_features`, which could be a user-defined function that modifies stuff. mypy can only infer that `__getitem__` of `List[T]` returns `T`, and nothing else.

If you store `self._dataset_features[i]` in a local variable and do a `is not None` check on it, ignores might not be needed anymore.